src/magnetVacuumFF.cpp

```diff
@@ -31,155 +24,256 @@ using namespace dd4hep;
-  pv_b0pf_vacuum.addPhysVolID("sector", 1);
-  DetElement tube_de_1(sdet, "sector1_de", 1);
-  tube_de_1.setPlacement(pv_b0pf_vacuum);
-
-  auto pv_b0apf_vacuum =
-      assembly.placeVolume(v_b0apf_vacuum, Transform3D(RotationY(rotation_b0apf), Position(x_b0apf, y_b0apf, z_b0apf)));
-  pv_b0apf_vacuum.addPhysVolID("sector", 1);
-  DetElement tube_de_2(sdet, "sector2_de", 1);
-  tube_de_2.setPlacement(pv_b0apf_vacuum);
-
-  auto pv_q1apf_vacuum =
-      assembly.placeVolume(v_q1apf_vacuum, Transform3D(RotationY(rotation_q1apf), Position(x_q1apf, y_q1apf, z_q1apf)));
-  pv_q1apf_vacuum.addPhysVolID("sector", 1);
-  DetElement tube_de_3(sdet, "sector3_de", 1);
-  tube_de_3.setPlacement(pv_q1apf_vacuum);
-
-  auto pv_q1bpf_vacuum =
-      assembly.placeVolume(v_q1bpf_vacuum, Transform3D(RotationY(rotation_q1bpf), Position(x_q1bpf, y_q1bpf, z_q1bpf)));
-  pv_q1bpf_vacuum.addPhysVolID("sector", 1);
-  DetElement tube_de_4(sdet, "sector4_de", 1);
-  tube_de_4.setPlacement(pv_q1bpf_vacuum);
-
-  auto pv_q2pf_vacuum =
-      assembly.placeVolume(v_q2pf_vacuum, Transform3D(RotationY(rotation_q2pf), Position(x_q2pf, y_q2pf, z_q2pf)));
-  pv_q2pf_vacuum.addPhysVolID("sector", 1);
-  DetElement tube_de_5(sdet, "sector5_de", 1);
-  tube_de_5.setPlacement(pv_q2pf_vacuum);
-
-  auto pv_b1pf_vacuum =
-      assembly.placeVolume(v_b1pf_vacuum, Transform3D(RotationY(rotation_b1pf), Position(x_b1pf, y_b1pf, z_b1pf)));
-  pv_b1pf_vacuum.addPhysVolID("sector", 1);
-  DetElement tube_de_6(sdet, "sector6_de", 1);
-  tube_de_6.setPlacement(pv_q1apf_vacuum);
-
-  auto pv_b1apf_vacuum =
-      assembly.placeVolume(v_b1apf_vacuum, Transform3D(RotationY(rotation_b1apf), Position(x_b1apf, y_b1apf, z_b1apf)));
-  pv_b1apf_vacuum.addPhysVolID("sector", 1);
-  DetElement tube_de_7(sdet, "sector7_de", 1);
-  tube_de_7.setPlacement(pv_b1apf_vacuum);
-
-
-  pv_assembly = det.pickMotherVolume(sdet).placeVolume(assembly); //, posAndRot);
-  pv_assembly.addPhysVolID("system", x_det.id()).addPhysVolID("barrel", 1);
-  sdet.setPlacement(pv_assembly);
-  assembly->GetShape()->ComputeBBox();
-  return sdet;
+        using namespace ROOT::Math;
+        xml_det_t x_det    = e;
+        string    det_name = x_det.nameStr();
+        DetElement sdet(det_name, x_det.id());
+        Assembly   assembly(det_name + "_assembly");
+        Material   m_Vac    = det.material("Vacuum");
+        string     vis_name = x_det.visStr();
+
+        PlacedVolume pv_assembly;
+
+        //----------------------------------------------
+        // Starting point is only the magnet centers,
+        // lengths, rotations, and radii -->
+        // everything else calculated internally to
+        // make it easier to update later.
+        //----------------------------------------------
+
+        const int numGaps = 6; //number of gaps between magnets (excluding the IP to B0pf transition -- special case)
```

Comment:
Instead of using a prefilled dimension, I'd suggest a dynamically expanding vector of struct with radius etc. Then you can fill this in the loop, and avoid the index idx.

Reply:
This is fixed - everything is std::vectors now.